File: backend/easyfold/af3_input/_chain_ids.py
"""Excel-column-style chain ID generator (1-indexed)."""


def excel_chain_id(n: int) -> str:
    """Return the n-th chain ID in Excel column order.

    1 → "A", 26 → "Z", 27 → "AA", 28 → "AB", …, 52 → "AZ",
    53 → "BA", …, 702 → "ZZ", 703 → "AAA".

    Args:
        n: 1-based index. Must be >= 1.

    Raises:
        ValueError: when ``n < 1``.
    """
    if n < 1:
        raise ValueError(f"chain id index must be >= 1, got {n}")
    result = ""
    remaining = n
    while remaining > 0:
        remaining, rem = divmod(remaining - 1, 26)
        result = chr(ord("A") + rem) + result
    return result

File: frontend/components/assembly-builder/chain-ids.ts
/**
 * Excel-column-style chain ID generator (1-indexed).
 *
 * Mirrors `backend/easyfold/af3_input/_chain_ids.py` so the assembly preview
 * shows the same chain IDs the backend will assign at job-build time. Keep
 * the two implementations in sync — the backend is the source of truth for
 * chain ordering, this is purely a UX preview.
 */

export function excelChainId(n: number): string {
  if (n < 1) throw new Error(`chain id index must be >= 1, got ${n}`);
  let result = "";
  let remaining = n;
  while (remaining > 0) {
    const rem = (remaining - 1) % 26;
    remaining = Math.floor((remaining - 1) / 26);
    result = String.fromCharCode("A".charCodeAt(0) + rem) + result;
  }
  return result;
}

/** Returns a comma-separated string of chain IDs starting at `startIndex` for `count` chains. */
export function chainIdRange(startIndex: number, count: number): string {
  const ids: string[] = [];
  for (let i = 0; i < count; i++) {
    ids.push(excelChainId(startIndex + i));
  }
  return ids.join(", ");
}
